File: src/ch/jbead/RectIterator.java
/** jbead - http://www.brunoldsoftware.ch
    Copyright (C) 2001-2012  Damian Brunold

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

package ch.jbead;

import java.util.Iterator;

public class RectIterator implements Iterator<Point> {

    private Point begin;
    private Point end;
    private Point next;
    
    public RectIterator(Point begin, Point end) {
        this.begin = new Point(Math.min(begin.getX(), end.getX()), Math.min(begin.getY(), end.getY()));
        this.end = new Point(Math.max(begin.getX(), end.getX()), Math.max(begin.getY(), end.getY()));
        this.next = this.begin;
    }
    
    @Override
    public boolean hasNext() {
        return next != null;
    }

    @Override
    public Point next() {
        Point result = next;
        next = next.nextRight();
        if (next.getX() > end.getX()) {
            next.setX(begin.getX());
            next.setY(next.getY() + 1);
        }
        if (next.getY() > end.getY()) {
            next = null;
        }
        return result;
    }

    @Override
    public void remove() {
        // ignore
    }

}

File: csharp/JBead.Web/Core/RectIterator.cs
using System.Collections;

namespace JBead.Web.Core;

public sealed class RectIterator : IEnumerator<Point>
{
    private readonly Point begin;
    private readonly Point end;
    private Point? next;
    private Point current;

    public RectIterator(Point begin, Point end)
    {
        this.begin = new Point(Math.Min(begin.X, end.X), Math.Min(begin.Y, end.Y));
        this.end = new Point(Math.Max(begin.X, end.X), Math.Max(begin.Y, end.Y));
        next = this.begin.X < 0 || this.end.Y < 0 ? null : this.begin;
    }

    public Point Current => current;
    object IEnumerator.Current => current;

    public bool MoveNext()
    {
        if (next is null) {
			return false;
		}
		current = next.Value;
        var advanced = current.NextRight();
        if (advanced.X > end.X) {
			advanced = new Point(begin.X, advanced.Y + 1);
		}
		next = advanced.Y > end.Y ? null : advanced;
        return true;
    }

    public void Reset()
    {
        next = begin.X < 0 || end.Y < 0 ? null : begin;
    }

    public void Dispose() { }
}
